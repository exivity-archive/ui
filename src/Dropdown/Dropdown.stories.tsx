import React from 'react'
// @ts-ignore
import { storiesOf } from '@storybook/react'
import { withState } from '../utils/tests/decorators/StateDecorator'
import { Dropdown } from '.'

storiesOf('molecules|Dropdown', module)
  .addDecorator(withState(false))
  .add('default', ({ state, storeState }: any) => {
    return (
      <div style={{ padding: '300px 0' }}>
        <Dropdown
          open={state}
          button={
            <div onClick={() => storeState(!state)}>Click me</div>
          }>
          <div style={{ width: 500 }}>
            Lorem Ipsum is simply dummy text of the printing and typesetting industry.
            Lorem Ipsum has been the industry's standard dummy text ever since the 1500s,
            when an unknown printer took a galley of type and scrambled it to make a type specimen book.
            It has survived not only five centuries,
            but also the leap into electronic typesetting, remaining essentially unchanged.
            It was popularised in the 1960s with the release of Letraset sheets containing Lorem Ipsum passages,
            and more recently with desktop publishing software like Aldus PageMaker including versions of Lorem Ipsum
          </div>
        </Dropdown>
      </div>
    )
  })
