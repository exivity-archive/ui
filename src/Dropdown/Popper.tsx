import React from 'react'
import { Manager, Reference, Popper as ReactPopper } from 'react-popper'
import styled from 'styled-components'

import { OutsideClickListener } from '../OutsideClickListener'

export const Container = styled.div`
  position: relative;
`

/**
 * Placement options for Popper
 * @see https://popper.js.org/popper-documentation.html#Popper.placements
 */
export enum Placement {
  RIGHT = 'right',
  RIGHT_START = 'right-start',
  RIGHT_END = 'right-end',
  LEFT = 'left',
  LEFT_START = 'left-start',
  LEFT_END = 'left-end',
  TOP = 'top',
  TOP_START = 'top-start',
  TOP_END = 'top-end',
  BOTTOM = 'bottom',
  BOTTOM_START = 'bottom-start',
  BOTTOM_END = 'bottom-end',
  AUTO = 'auto',
  AUTO_START = 'auto-start',
  AUTO_END = 'auto-end'
}

interface TriggerProps {
  ref: React.Ref<any>
}

interface ChildProps {
  ref: React.Ref<HTMLElement>
  style: React.CSSProperties
  placement: Placement
  scheduleUpdate: () => void
  arrowProps: {
    ref: React.Ref<HTMLElement>
    style: React.CSSProperties
  }
}

export interface PopperProps {
  renderTrigger: (props: TriggerProps) => React.ReactNode
  open: boolean
  children: (props: ChildProps) => React.ReactNode
  placement?: Placement
  onOutsideClick?: (...rest: any) => void
  offset?: number
  flip?: boolean
}

export const Popper: React.FC<PopperProps> = ({
  renderTrigger,
  open,
  children,
  onOutsideClick,
  placement = Placement.BOTTOM_START,
  offset = 0,
  flip = true
}) => (
  <Manager>
    <Container>
      <OutsideClickListener onOutsideClick={onOutsideClick}>
        <Reference>{renderTrigger}</Reference>
        {open
          ? (
            <ReactPopper
              placement={placement}
              positionFixed={false}
              modifiers={{
                flip: { enabled: flip },
                offset: { enabled: offset !== 0, offset: `${offset}, ${offset}` }
              }}>
              {({ ref, style, placement: currentPlacement, arrowProps, scheduleUpdate }) =>
                children({
                  ref: ref as React.Ref<HTMLElement>,
                  style,
                  placement: currentPlacement as Placement,
                  arrowProps,
                  scheduleUpdate
                })
              }
            </ReactPopper>
          )
          : null
        }
      </OutsideClickListener>
    </Container>
  </Manager>
)
