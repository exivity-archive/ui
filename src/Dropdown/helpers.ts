export interface Rects {
  outer: ClientRect | DOMRect
  inner: ClientRect | DOMRect
}

export type Vertical = 'top' | 'bottom' | 'auto'
export type Horizontal = 'left' | 'right' | 'auto'

export interface Layout {
  horizontal: Horizontal
  vertical: Vertical,
}

export const elementCrossedEdge = (absolutePosition: number, elementDimension: number, edge: number) => {
  return absolutePosition + elementDimension > edge
}

export function getVertical (vertical: Vertical, { outer, inner }: Rects, breakDistance: number) {
  if (vertical === 'auto') {
    return elementCrossedEdge(outer.bottom, inner.height, window.innerHeight - breakDistance)
      ? 'top' : 'bottom'
  } else {
    return vertical
  }
}

export function getHorizontal (horizontal: Horizontal, { outer, inner }: Rects, breakDistance: number) {
  if (horizontal === 'auto') {
    return elementCrossedEdge(outer.left, inner.width, window.innerWidth - breakDistance)
      ? 'left' : 'right'
  } else {
    return horizontal
  }
}

export interface Position {
  left?: 0
  right?: 0
  top?: number
  bottom?: number
}

export function getPosition (rects: Rects, layout: Layout, breakDistance: number): Position {
  const vertical = getVertical(layout.vertical, rects, breakDistance)
  const horizontal = getHorizontal(layout.horizontal, rects, breakDistance)

  return {
    [vertical]: rects.outer.height,
    [horizontal]: 0
  }
}

export function makeCssPosition ({ top, right, bottom, left }: Position) {
  let position = ''
  if (top !== undefined) position += `bottom: ${top}px;\n` + 'margin-top: 5px;\n'
  if (right !== undefined) position += `left: ${right}px;\n`
  if (bottom !== undefined) position += `top: ${bottom}px;\n` + 'margin-bottom: 5px;\n'
  if (left !== undefined) position += `right: ${left}px;\n`
  return position
}
