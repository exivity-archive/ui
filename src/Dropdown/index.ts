export * from './Dropdown'
export * from './useClosable'
