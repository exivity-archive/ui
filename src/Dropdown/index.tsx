import React, { useRef, useState, useEffect, useLayoutEffect } from 'react'
import styled, { css } from 'styled-components'

interface IPosition {
  left?: 0
  right?: 0
  top?: number
  bottom?: number
}

interface IContentProps extends IPosition {
  opened: boolean
}

const Content = styled.div<IContentProps>`
  position: absolute;
  background-color: #f9f9f9;
  min-width: 160px;
  box-shadow: 0 5px 15px rgba(0,0,0,0.08);
  padding: 20px;
  padding-right: 10%;
  visibility: ${({ opened }) => opened ? 'visible' : 'hidden'};
  ${ ({ left, right, top, bottom }) => css`
    ${typeof left !== 'undefined' && css`
      left: ${left};
    `}
    ${typeof right !== 'undefined' && css`
      right: ${right};
    `}
    ${typeof top !== 'undefined' && css`
      top: ${top}px;
      margin-top: 5px;
    `}
    ${typeof bottom !== 'undefined' && css`
      bottom: ${bottom}px;
      margin-bottom: 5px;
    `}
  `}
`

type Vertical = 'to-top' | 'to-bottom'
type Horizontal = 'to-left' | 'to-right'
interface ILayout {
  horizontal: Horizontal
  vertical: Vertical,
}
interface IDropdownProps {
  className?: string
  button: React.ReactNode
  opened: boolean
  layout?: ILayout | 'auto'
}


const Dropdown: React.FC<IDropdownProps> = ({ 
  className, 
  button, 
  children, 
  opened, 
  layout = 'auto'
}) => {
  const ref = useRef<HTMLDivElement>(null)
  const [position, setPosition] = useState<IPosition | undefined>()

  const getPosition = (layout: ILayout | 'auto') => {
    if (layout !== 'auto' && position !== undefined) {
      return position
    } else if (ref.current) {
      const rect = ref.current.getBoundingClientRect()
      const parentRect = ref.current.parentElement!.getBoundingClientRect()
      
      const makePosition = ({ horizontal, vertical }: ILayout): IPosition => ({
        [horizontal === 'to-left' ? 'right' : 'left']: 0,
        [vertical === 'to-top' ? 'bottom' : 'top']: parentRect.height
      })

      if (layout === 'auto') {
        const breakDistance = 20
        const closeToBottomBound = parentRect.bottom + rect.height > window.innerHeight - breakDistance
        const closeToRightBound = parentRect.left + rect.width > window.innerWidth - breakDistance 

        return makePosition({
          vertical: closeToBottomBound ? 'to-top' : 'to-bottom', 
          horizontal: closeToRightBound ? 'to-left' : 'to-right'
        })
      } else {
        return makePosition(layout)

      }
    }
  }

  const handlePosition = () => setPosition(getPosition(layout))

  useEffect(() => {
    window.addEventListener('resize', handlePosition)  
    return () => window.removeEventListener('resize', handlePosition)  
  }, [])

  useLayoutEffect(handlePosition, [])

  return (
    <div className={className} data-test='dropdown'>
      {button}
        <Content 
          data-test='dropdown-content'
          ref={ref}
          {...position}
          opened={opened}>

          {children}
        </ Content>
    </div>
  )
}

export default styled(Dropdown)`
  position: relative;
`
