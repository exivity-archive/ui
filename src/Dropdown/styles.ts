import styled from 'styled-components'
import { Block } from '../Block'
import { fromTheme } from '../utils'

export const Content = styled(Block)<{ fullWidth?: boolean }>`
  background-color: ${fromTheme(t => t.colors.white)};
  box-shadow: 0 5px 15px rgba(0,0,0,0.08);
  min-width: ${({ fullWidth }) => fullWidth ? '100%' : '160px'};
  z-index: ${fromTheme(theme => theme.global.zPriority.foreground)};
`
