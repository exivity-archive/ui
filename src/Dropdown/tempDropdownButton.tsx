import * as React from 'react'

import styled from 'styled-components'

const SelectedValue = styled.div`
  padding-top: 3px;
  float: left;
`

interface IDropdownButtonProps {
  value?: string
  className?: string
  onClick?: () => void
}

const DropdownButton: React.FC<IDropdownButtonProps> = ({ value, className, onClick }) => (
  <button className={className} onClick={onClick}>
    <SelectedValue >{value}</SelectedValue>
  </button>
)

export default styled(DropdownButton)`
  font-family: inherit;
  color: #444444;
  height: 30px;
  width: 330px;
  background-color: #F4F4F4;
  border: none;
  outline: none;
  padding: 5px 15px;
  cursor: pointer;
`
