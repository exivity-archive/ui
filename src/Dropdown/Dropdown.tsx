import React, { useEffect, useLayoutEffect, useRef, useState, useMemo } from 'react'
import styled, { css } from 'styled-components'

import { OutsideClickListener } from '../OutsideClickListener'
import { Block, BlockProps } from '../Block'

import {
  makeCssPosition
} from './helpers'

import { fromTheme } from '../utils/styled'
import { useEdgeAvoidingLayout } from '../useEdgeAvoidingLayout'
import { Vertical, Horizontal } from '../useEdgeAvoidingLayout/helpers'

const StyledDropdown = styled.div`
  position: relative;
  width: 100%;
`

interface ContentProps {
  open: boolean
  useTriggerComponentWidth?: boolean
  width?: string
  position: string
}

const Content = styled.div <ContentProps>`
  box-sizing: border-box;
  position: absolute;
  background-color: #f9f9f9;
  box-shadow: 0 5px 15px rgba(0,0,0,0.08);
  z-index: ${fromTheme(theme => theme.global.zPriority.one)};

  min-width: 160px;
  ${props => (props.useTriggerComponentWidth && props.width) && css`
    width: ${props.width};
  `}

  visibility: ${({ open }) => open ? 'visible' : 'hidden'};
  ${({ position }) => `${position}`}
`

export interface DropdownProps extends BlockProps {
  className?: string
  triggerComponent: React.ReactNode
  open: boolean
  vertical?: Vertical
  horizontal?: Horizontal
  breakDistance?: number
  useTriggerComponentWidth?: boolean
  onOutsideClick?: (...rest: any) => void
  test?: string
}

export const Dropdown: React.FC<DropdownProps> = ({
  className,
  triggerComponent,
  children,
  open,
  horizontal = 'auto',
  vertical = 'auto',
  breakDistance = 20,
  useTriggerComponentWidth,
  onOutsideClick,
  test = 'dropdown',
  ...blockProps
}) => {
  const [refs, handleLayout, layout] = useEdgeAvoidingLayout<HTMLDivElement, HTMLDivElement>(
    breakDistance, { horizontal, vertical }
  )

  useEffect(() => {
    window.addEventListener('resize', handleLayout)
    return () => window.removeEventListener('resize', handleLayout)
  }, [])

  useLayoutEffect(handleLayout, [])

  const triggerWidth = refs.parent.current
    ? `${refs.parent.current.clientWidth}px`
    : undefined

  const position = useMemo(() => {
    if (refs.parent.current) {
      return makeCssPosition(layout, refs.parent.current.getBoundingClientRect().height)
    }
    return ''
  }, [layout.vertical, layout.horizontal])

  return (
    <StyledDropdown className={className} data-test={test} ref={refs.parent}>
      <OutsideClickListener onOutsideClick={onOutsideClick}>
        {triggerComponent}
        <Content useTriggerComponentWidth={useTriggerComponentWidth}
          width={useTriggerComponentWidth ? triggerWidth : undefined}
          data-test={`${test}-content`}
          ref={refs.target}
          position={position}
          open={open}>
          <Block {...blockProps}>
            {children}
          </Block>
        </Content>
      </OutsideClickListener>
    </StyledDropdown>
  )
}
