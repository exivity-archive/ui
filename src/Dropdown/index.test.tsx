import * as React from 'react'
import { shallow, mount } from 'enzyme'
import 'jest-styled-components'

import Dropdown from '.'

//  @todo test position
test('renders dropdown', () => {
  const checkbox = shallow(<Dropdown open button={<div>test</div>}>test</Dropdown>)
  expect(checkbox).toMatchSnapshot()
})

test('passing checked prop as true should show dropdown content', () => {
  const dropdown = mount(<Dropdown open button={<div>test</div>}>test</Dropdown>)

  const dropdownContent = dropdown.find({ 'data-test': 'dropdown-content' })

  expect(dropdownContent).toHaveStyleRule('visibility', 'visible')
  dropdown.unmount()
})

test('passing checked prop as false should hide dropdown content', () => {
  const dropdown = mount(<Dropdown open={false} button={<div>test</div>}>test</Dropdown>)

  const dropdownContent = dropdown.find({ 'data-test': 'dropdown-content' })

  expect(dropdownContent).toHaveStyleRule('visibility', 'hidden')
  dropdown.unmount()
})
