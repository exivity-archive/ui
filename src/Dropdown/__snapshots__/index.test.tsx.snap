// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders dropdown 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ForwardRef
    button={
      <div>
        test
      </div>
    }
    opened={true}
  >
    test
  </ForwardRef>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "button": <div>
        test
      </div>,
      "children": "test",
      "forwardedComponent": Object {
        "$$typeof": Symbol(react.forward_ref),
        "attrs": Array [],
        "componentStyle": ComponentStyle {
          "componentId": "sc-bwzfXH",
          "isStatic": true,
          "rules": Array [
            "
  position: relative;
",
          ],
        },
        "displayName": "Styled(Dropdown)",
        "foldedComponentIds": Array [],
        "render": [Function],
        "styledComponentId": "sc-bwzfXH",
        "target": [Function],
        "toString": [Function],
        "warnTooManyClasses": [Function],
        "withComponent": [Function],
        Symbol(Symbol.hasInstance): [Function],
      },
      "forwardedRef": null,
      "opened": true,
    },
    "ref": null,
    "rendered": "test",
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "button": <div>
          test
        </div>,
        "children": "test",
        "forwardedComponent": Object {
          "$$typeof": Symbol(react.forward_ref),
          "attrs": Array [],
          "componentStyle": ComponentStyle {
            "componentId": "sc-bwzfXH",
            "isStatic": true,
            "rules": Array [
              "
  position: relative;
",
            ],
          },
          "displayName": "Styled(Dropdown)",
          "foldedComponentIds": Array [],
          "render": [Function],
          "styledComponentId": "sc-bwzfXH",
          "target": [Function],
          "toString": [Function],
          "warnTooManyClasses": [Function],
          "withComponent": [Function],
          Symbol(Symbol.hasInstance): [Function],
        },
        "forwardedRef": null,
        "opened": true,
      },
      "ref": null,
      "rendered": "test",
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
